Improve Feedback path UI #5604

YongJieYongJie opened the pull request:
Fixes #5604 

Here's how it looks:
_(I've tried not to change the wording as much as possible; leaving it to another issue)_
![normal](https://cloud.githubusercontent.com/assets/14101781/16404071/1c210682-3d2e-11e6-9a92-8915bad38d53.png)
![expanded](https://cloud.githubusercontent.com/assets/14101781/16404073/1eb1ec4a-3d2e-11e6-83a4-eb916f364538.png)
![other](https://cloud.githubusercontent.com/assets/14101781/16404074/20f57ac6-3d2e-11e6-9c59-bd166d565047.png)

damithc commented:
Users may not get the difference between the last two drop-down options.
`Other options` --> `More combinations like the ones above`? Not sure if this is better though.

YongJieYongJie commented:
@damithc What about something like this:

![dropdown header](https://cloud.githubusercontent.com/assets/14101781/16439776/22eec86c-3dee-11e6-9054-7ce1fdab18aa.png)

damithc commented:
Yes, this one looks better.

YongJieYongJie commented:
Removed the option for 'Custom feedback path' (can be added back once that feature is complete)

![removed custom](https://cloud.githubusercontent.com/assets/14101781/16550360/3a6616a2-41dc-11e6-951b-5447eb52e072.png)

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java:
```diff
@@ -101,4 +127,18 @@ public String toSingularFormString() {
     }
+
+    /**
+     * Getter method for JSTL to access the {@code displayNameGiver} property directly
```
You don't need to mention "for JSTL". In fact, this can be as simple as "Gets the ABC property".

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java:
```diff
@@ -17,6 +21,28 @@
+                giverInitializer.add(participantType);
+            }
+        }
+        GIVERS = Collections.unmodifiableList(giverInitializer);
```
What's the purpose of the variables `GIVERS` and `RECIPIENTS`?

YongJieYongJie replied:
It is used to directly generate the `<option>` for feedback path:
https://github.com/TEAMMATES/teammates/blob/fd9dfb48da8508bbc7fdf8058d3f9ac14f9114e7/src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/questionFeedbackPathSettings.tag#L64

wkurniawan07 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -418,6 +419,36 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
+                                                  FeedbackParticipantType.SELF,
+                                                  FeedbackParticipantType.INSTRUCTORS)));
+
+            COMMON_FEEDBACK_PATHS = initializer;
```
Instead of re-assigning, won't `COMMON_FEEDBACK_PATHS.put(...)` work?

YongJieYongJie replied:
Oh, I made a mistake here, I'm supposed to re-assign while using `Collections.unmodifiableMap(initializer)` to create the const

wkurniawan07 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -418,6 +419,36 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
     
     public static class FeedbackQuestion {
+
+        public static final LinkedHashMap<FeedbackParticipantType, List<FeedbackParticipantType>>
```
Let's keep this one as `Map`. Similarly for the `initializer` variable.

wkurniawan07 commented on src/main/java/teammates/ui/template/FeedbackQuestionFeedbackPathSettings.java:
```diff
@@ -48,4 +35,27 @@ public void setNumOfEntitiesToGiveFeedbackToValue(int numOfEntitiesToGiveFeedbac
         this.numOfEntitiesToGiveFeedbackToValue = numOfEntitiesToGiveFeedbackToValue;
     }
     
+    public boolean getIsCommonPathSelected() {
```
This one would be better as `isCommonPathSelected()`.

YongJieYongJie replied:
I actually wanted the JSTL to read nicer: `.isCommonPathSelected` as opposed to `.commonPathSelected`, which can be slightly misleading

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/questionFeedbackPathSettings.tag:
```diff
@@ -5,74 +5,113 @@
-                    </option>
-                </c:forEach>
-            </select>
+    <div class="feedback-path-others margin-top-7px"<c:if test="${fqForm.feedbackPathSettings.isCommonPathSelected || isNewQuestion}"> style="display:none;"</c:if>">
```
Extraneous `"` at the end.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageData.java:
```diff
@@ -148,6 +140,28 @@ private void buildExistingQuestionForm(String feedbackSessionName,
+        settings.setNumberOfEntitiesToGiveFeedbackToChecked(isNumberOfEntitiesToGiveFeedbackToChecked);
+        settings.setNumOfEntitiesToGiveFeedbackToValue(isNumberOfEntitiesToGiveFeedbackToChecked
+                                                           ? question.numberOfEntitiesToGiveFeedbackTo
+                                                           : 1);
```
This is better as it was before, i.e don't add another level of indentation for the `?` and `:`.

wkurniawan07 commented on src/main/webapp/js/instructorFeedbackEdit/feedbackPath.js:
```diff
@@ -0,0 +1,70 @@
+$(document).ready(function() {
+    FeedbackPath.attachEvents();
+});
```
Can this be moved to the js file specifically for the instructor feedback edit page?
That way, we can unit test the `FeedbackPath` variable on its own without having to worry about the document.ready even being called. I actually want to implement this kind of js paradigm across the code base in the near future.

wkurniawan07 commented on src/main/webapp/js/instructorFeedbackEdit/feedbackPath.js:
```diff
@@ -0,0 +1,70 @@
+var FeedbackPath = {
+    attachEvents: function() {
+        var allDropdownOptions = $('.feedback-path-dropdown-option');
+        FeedbackPath.attachEventsForAllOptions(allDropdownOptions);
```
Is the "class" name `FeedbackPath` needed? In this closure just the function name should be enough, right?

YongJieYongJie replied:
We'll need at least a `this`, I chose `FeedbackPath` for clarify, and because we are not doing OOP

wkurniawan07 commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -735,6 +735,58 @@ Apply to thead to make headers not bold.
+//////////////////////////////////////////////////////////*/
+
+.dropdown-submenu {
+    position:relative;
```
Add a space after `:`. Applies to all the new css rules added here.

wkurniawan07 commented:
Added some comments to work on. I'll have another review afterwards.

YongJieYongJie commented:
@wkurniawan07 I fixed the preliminary errors.

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/questionFeedbackPathSettings.tag:
```diff
@@ -5,74 +5,113 @@
+            <c:forEach items="<%= Const.FeedbackQuestion.COMMON_FEEDBACK_PATHS %>" var="commonPath">
+                <li class="dropdown-submenu">
+                    <c:set var="commonGiver" value="${commonPath.key}" />
+                    <a>${commonGiver.displayNameGiver} will give feedback on...</a>
```
"... will give feedback ~~on~~to..."

YongJieYongJie replied:
I think actually since the recipient may not be able to see the response, we should use `on` instead of `to`

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/questionFeedbackPathSettings.tag:
```diff
@@ -5,74 +5,113 @@
+                            <c:forEach items="${commonPath.value}" var="commonRecipient">
+                                <a class="feedback-path-dropdown-option" href="javascript:;"
+                                   data-giver-type="${commonGiver}" data-recipient-type="${commonRecipient}"
+                                   data-path-description="${commonGiver.displayNameGiver} will give feedback on <span class='glyphicon glyphicon-arrow-right'></span> ${commonRecipient.displayNameRecipient}">
```
Similarly here.

wkurniawan07 commented on src/main/webapp/js/instructorFeedbackEdit/feedbackPath.js:
```diff
@@ -0,0 +1,66 @@
+        FeedbackPath.attachEventsForAllOptions(allDropdownOptions);
+
+        var commonOptions = allDropdownOptions.not(
+                '.feedback-path-dropdown-option-other');
```
This is fine on the previous line.

wkurniawan07 commented:
Looks good, other than the few suggested improvements :+1:

YongJieYongJie commented:
@wkurniawan07 Changes made, except the `to/on` issue.

wkurniawan07 commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -735,6 +735,58 @@ Apply to thead to make headers not bold.
+    position: relative;
+}
+
+.dropdown-submenu>.dropdown-menu {
```
Add a space around the `>`; applies to the other new classes added as well.

wkurniawan07 commented:
Looks good other than the minor change

damithc commented:
Help page updated to match?

YongJieYongJie commented:
@damithc Not yet, shall proceed to work on it

YongJieYongJie commented:
@wkurniawan07 Ready for review!
I copied the html source for the new feedback path div and replaced the old one in `instructorHelp.jsp`.
I also added `<script>` to the `helpPage.tag` so the interactions with the feedback path dropdown works.

![2016-07-15-155954_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/16867591/39c28b32-4aa5-11e6-8112-1c6212f0b321.png)

The showing/hiding of number of entities box wasn't working previously and I did not fix it because the fix is non trivial: unlike `visibilityOptions.js` and `feedbackPath.js` which are sort of modular-ish and works simply by including the js file and calling the attach event function, the code for showing/hiding the number of entities box is quite tightly bound in `instructorFeedbackEdit.js` and including it
 will require including several other possibly unrelated files)

wkurniawan07 commented:
@YongJieYongJie I don't recall `instructorHelp.jsp` having interactions enabled, i.e as far as I know the rendering of features are simply for presentation. Do you think it's a good idea to enable interaction for this feature only? Of course I could be wrong.

wkurniawan07 commented on src/main/webapp/instructorHelp.jsp:
```diff
@@ -1914,62 +1914,147 @@
-                                        </div>
+                                    <div class="feedback-path-dropdown btn-group">
+                                        <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                                            Predefined combinations:
```
This is the very reason why I call `instructorHelp.jsp` a "dreadful page" -- we have to hardcode everything that can be rendered with tags. But until we go to that direction, we don't really have another choice here.

damithc commented:
Interactions should not be enabled as far as possible. These are like screenshots but rendered using code.

YongJieYongJie commented:
@wkurniawan07 @damithc noted that interactions should not be enabled, will make the necessary changes (i.e., choose a static state with most options visible)

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java:
```diff
@@ -759,6 +759,24 @@ public void selectRecipientsToBeStudents(int qnNumber) {
+        WebElement otherOption = questionTable.findElement(
+                                     By.className("feedback-path-dropdown-option-other"));
+        dropdownButton.click();
+        otherOption.click();
```
I saw the Travis build report. All these `*.click()` must be `click(*)` as per the new version of Selenium's change in implementation.

wkurniawan07 commented:
@YongJieYongJie is this ready for review?

YongJieYongJie commented:
@wkurniawan07 Sorry for the delay, has been rather sick this whole week. I'll look through the changes when I get better and then set it to review.

wkurniawan07 commented:
@YongJieYongJie is this ready for review?

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java:
```diff
@@ -273,7 +275,8 @@ private void testEditQuestionAction() throws Exception {
         feedbackEditPage.clickVisibilityOptionsForQuestion1();
-        feedbackEditPage.selectGiverTypeForQuestion1("Me (Session creator)");
+        feedbackEditPage.enableOtherFeedbackPathOptions(1);
+        feedbackEditPage.selectGiverTypeForQuestion1("Feedback session creator (i.e., me)");
```
I'm quite sure you've migrated all `*ForQuestion1(...)` functions to something else, right?

YongJieYongJie replied:
It''s actually in another pull request that is current on hold #5802

wkurniawan07 commented on src/main/webapp/js/instructorHelp.js:
```diff
@@ -0,0 +1,3 @@
+$(document).ready(function() {
+    FeedbackPath.attachEvents();
```
This file should not be needed anymore

wkurniawan07 commented:
Some minor changes, everything else looks good!

YongJieYongJie commented:
@wkurniawan07 Ready for review!

wkurniawan07 commented:
Looks good! :+1:

damithc commented:
@chowyb @kanghj This is a major change in a critical feature, so we should do some additional manual testing before merging. Can help?

kanghj commented:
Got it, will manually test it

kanghj commented on src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/questionFeedbackPathSettings.tag:
```diff
@@ -5,47 +5,86 @@
+        <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+            <c:choose>
+                <c:when test="${isNewQuestion}">
+                    Please select a feedback path <span class="caret"></span>
```
This never seems to appear even when adding a new question?

YongJieYongJie replied:
The variable was declared wrongly, should be as follows:

```
-<c:set var="isNewQuestion" value="${empty fqForm.questionIndex}" />
+<c:set var="isNewQuestion" value="${fqForm.questionIndex eq -1}" />
```

chowyb commented:
Just saw this, manually testing it now.

chowyb commented:
Common feedback path combinations do not seem to be automatically selected if the new question's initial feedback path (which is copied over from the last question in the session) corresponds to one of the common paths, but that could be an intentional design choice. Haven't found any bugs otherwise.

YongJieYongJie commented:
@chowyb I added the behaviour for new question to check if a common path is selected.
@kanghj I fixed the `isNewQuestion` variable.

kanghj commented:
Looks good to me

chowyb commented:
Seems like if a non-common path is the initial feedback path, the selection properly shows "Predefined combinations", but no other inputs are present:

![fpq](https://cloud.githubusercontent.com/assets/7546744/17203292/26d7535a-54d1-11e6-9258-379b6fa5d07f.png)

damithc commented:
Changing status back as there seems to be some small issues to resolve.

YongJieYongJie commented:
@chowyb I can't reproduce the error, can you give the exact steps?

chowyb commented:
@YongJieYongJie as discussed earlier, the steps to reproduce this are as follows:

1. Create a new feedback session, with "Session with your own questions" selected.
2. Add a new feedback question, with a feedback path from "Other predefined combinations...". Pick a feedback path not covered by the common feedback paths (I used "Teams in this course" -> "Other teams in the course" in this case).
3. Save the question.
4. Click on "Add New Question".

YongJieYongJie commented:
@chowyb Can you help check if the "predefined combinations:" issue is fixed?

chowyb commented:
@YongJieYongJie yep the issue has been fixed :+1:

wkurniawan07 commented:
Looks good, tried some of the functions myself and so far no apparent defects :+1:

damithc commented:
BTW, instructorHelp page is using 4-space indentations. I thought we are using 2 spaces for html?

wkurniawan07 commented:
As I mentioned back in #5877, the fact is we still use 4 spaces for most (if not all) HTML/JSP/tag files.

damithc commented:
I see. Better switch to 2 spaces for html-like files to avoid wasting space. May be do it opportunistically, starting with the files where it is in fact a problem?
